tools/pythonpkg/src/python_udf.cpp

```diff
@@ -0,0 +1,362 @@
+		for (auto &param : params) {
+			auto type = py::cast<shared_ptr<DuckDBPyType>>(param);
+			parameters[i++] = type->Type();
+		}
+	}
+
+	void AnalyzeSignature(const py::object &udf) {
+		auto signature_func = py::module_::import("inspect").attr("signature");
+		auto signature = signature_func(udf);
+		auto sig_params = signature.attr("parameters");
+		auto return_annotation = signature.attr("return_annotation");
+		if (!py::none().is(return_annotation)) {
+			shared_ptr<DuckDBPyType> pytype;
+			if (py::try_cast<shared_ptr<DuckDBPyType>>(return_annotation, pytype)) {
+				return_type = pytype->Type();
+			}
+		}
+		param_count = py::len(sig_params);
+		parameters.reserve(param_count);
+		auto params = py::dict(sig_params);
+		for (auto &item : params) {
+			auto &value = item.second;
+			shared_ptr<DuckDBPyType> pytype;
+			if (py::try_cast<shared_ptr<DuckDBPyType>>(value.attr("annotation"), pytype)) {
+				parameters.push_back(pytype->Type());
+			} else {
+				std::string kind = py::str(value.attr("kind"));
+				auto parameter_kind = ParameterKind::FromString(kind);
+				if (parameter_kind == ParameterKind::Type::VAR_POSITIONAL) {
+					varargs = LogicalType::ANY;
+				}
+				parameters.push_back(LogicalType::ANY);
+			}
+		}
+	}
+
+	ScalarFunction GetFunction() {
+		ScalarFunction scalar_function(name, std::move(parameters), return_type, func, nullptr, nullptr, nullptr,
+		                               nullptr, varargs, FunctionSideEffects::NO_SIDE_EFFECTS, null_handling);
+		return scalar_function;
+	}
+};
+
+} // namespace
+
+static py::list ConvertToSingleBatch(const string &timezone_config, vector<LogicalType> &types, vector<string> &names,
+                                     DataChunk &input) {
+	ArrowSchema schema;
+	ArrowConverter::ToArrowSchema(&schema, types, names, timezone_config);
+
+	py::list single_batch;
+	ArrowAppender appender(types, STANDARD_VECTOR_SIZE);
+	appender.Append(input, 0, input.size(), input.size());
+	auto array = appender.Finalize();
+	TransformDuckToArrowChunk(schema, array, single_batch);
+	return single_batch;
+}
+
+static py::object ConvertDataChunkToPyArrowTable(DataChunk &input, const string &timezone_config) {
+	py::gil_scoped_acquire acquire;
+
+	auto pyarrow_lib_module = py::module::import("pyarrow").attr("lib");
+	auto from_batches_func = pyarrow_lib_module.attr("Table").attr("from_batches");
+	auto schema_import_func = pyarrow_lib_module.attr("Schema").attr("_import_from_c");
+	ArrowSchema schema;
+
+	auto types = input.GetTypes();
+	vector<string> names;
+	names.reserve(types.size());
+	for (idx_t i = 0; i < types.size(); i++) {
+		names.push_back(StringUtil::Format("c%d", i));
+	}
+	ArrowConverter::ToArrowSchema(&schema, types, names, timezone_config);
+	auto schema_obj = schema_import_func((uint64_t)&schema);
+
+	py::list single_batch = ConvertToSingleBatch(timezone_config, types, names, input);
+
+	return py::cast<duckdb::pyarrow::Table>(from_batches_func(single_batch, schema_obj));
+}
```

Comment:
This seems rather similar to `duckdb::pyarrow::Table DuckDBPyResult::FetchArrowTable(idx_t rows_per_batch)` in `pyresult` maybe these can be unified in some ArrowExportUtils.

Reply:
I guess if I make the function take a callback to get the batches, that might be possible yea 👍